File: go/internal/store/likes.go
package store

import (
	"context"
	"fmt"

	"github.com/mudbase/mudbase-showcase-social/go/internal/mbase"
	"github.com/mudbase/mudbase-showcase-social/go/internal/models"
)

// likedPostIDsLimit bounds how many of the current visitor's own likes are fetched to mark hearts
// filled across a feed/profile page - generous enough that a real demo account never truncates.
const likedPostIDsLimit = 500

// LikeService implements every operation the app needs against the `likes` collection.
type LikeService struct {
	client       *mbase.Client
	collectionID string
	posts        *PostService
}

// NewLikeService builds a LikeService bound to the given likes collection ID. It needs *PostService
// too, since toggling a like also has to bump the post's denormalized likesCount.
func NewLikeService(client *mbase.Client, likesCollectionID string, posts *PostService) *LikeService {
	return &LikeService{client: client, collectionID: likesCollectionID, posts: posts}
}

// findLike looks up the (postID, userID) like row, ok=false if none exists.
func (s *LikeService) findLike(ctx context.Context, token, postID, userID string) (like models.Like, ok bool, err error) {
	result, err := mbase.List[models.Like](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"postId": postID, "userId": userID},
		Limit:  1,
	})
	if err != nil {
		return models.Like{}, false, fmt.Errorf("store: looking up like for post %s: %w", postID, err)
	}
	if len(result.Data) == 0 {
		return models.Like{}, false, nil
	}
	return result.Data[0], true, nil
}

// IsLiked reports whether userID currently likes postID.
func (s *LikeService) IsLiked(ctx context.Context, token, postID, userID string) (bool, error) {
	_, ok, err := s.findLike(ctx, token, postID, userID)
	return ok, err
}

// Toggle likes postID for userID if not already liked, or unlikes it if it is - a check-then-act
// query immediately before create/delete (see models.Like's doc comment on the tradeoff this
// implies). Either way, the post's denormalized likesCount is bumped to match. Returns the new
// liked state.
func (s *LikeService) Toggle(ctx context.Context, token, postID, userID string) (liked bool, err error) {
	existing, ok, err := s.findLike(ctx, token, postID, userID)
	if err != nil {
		return false, err
	}

	if ok {
		if err := mbase.Delete(ctx, s.client, token, s.collectionID, existing.ID); err != nil {
			return false, fmt.Errorf("store: removing like: %w", err)
		}
		if err := s.posts.BumpLikesCount(ctx, token, postID, -1); err != nil {
			return false, err
		}
		return false, nil
	}

	if _, err := mbase.Create[models.Like](ctx, s.client, token, s.collectionID, map[string]interface{}{
		"postId": postID,
		"userId": userID,
	}); err != nil {
		return false, fmt.Errorf("store: creating like: %w", err)
	}
	if err := s.posts.BumpLikesCount(ctx, token, postID, 1); err != nil {
		return false, err
	}
	return true, nil
}

// LikedPostIDs returns the set of post IDs userID currently likes, as a lookup map so feed/profile
// rendering can mark each post card's heart filled in O(1) per post - mirrors
// web/src/hooks/useLikes.ts's useMyLikedPostIds.
func (s *LikeService) LikedPostIDs(ctx context.Context, token, userID string) (map[string]bool, error) {
	result, err := mbase.List[models.Like](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"userId": userID},
		Limit:  likedPostIDsLimit,
	})
	if err != nil {
		return nil, fmt.Errorf("store: listing likes for user %s: %w", userID, err)
	}
	ids := make(map[string]bool, len(result.Data))
	for _, like := range result.Data {
		ids[like.PostID] = true
	}
	return ids, nil
}

File: python/app/services/likes.py
"""likes collection: one row per (postId, userId) pair. This collection type
has no compound unique index, so uniqueness is enforced at the application
layer — every toggle re-queries `{postId, userId}` immediately before
creating/deleting, a reasonable (not airtight) guard against a
double-click/double-tab race, matching the reference web app's
`useToggleLike`.
"""

import asyncio
import logging

from app.config import get_settings
from app.mudbase_client import MudbaseApiError, create_data_sync, delete_data_sync, list_data_sync
from app.schemas.social import Like
from app.services.posts import adjust_post_counts

logger = logging.getLogger(__name__)


async def find_like(post_id: str, user_id: str, *, access_token: str | None = None) -> Like | None:
    settings = get_settings()
    result = await asyncio.to_thread(
        list_data_sync,
        settings.likes_collection_id,
        filter_dict={"postId": post_id, "userId": user_id},
        limit=1,
        access_token=access_token,
    )
    docs = result["data"]
    return Like.model_validate(docs[0]) if docs else None


async def list_liked_post_ids(user_id: str, *, access_token: str) -> set[str]:
    """Used by the feed/profile pages to render each post's like button in
    its correct (already-liked or not) state for the current visitor.

    Capped at 100 rows (see the `limit` comment below) — a visitor with more
    than 100 likes on record will see the like button on some of their
    older-liked posts render as not-liked. This is an explicit, accepted
    tradeoff for a demo app (same precedent as `comments.py::list_comments`'s
    100-row cap), not an oversight: the like/comment counts on the post
    itself are unaffected either way, only this per-viewer overlay."""
    settings = get_settings()
    result = await asyncio.to_thread(
        list_data_sync,
        settings.likes_collection_id,
        filter_dict={"userId": user_id},
        limit=100,  # Mudbase's DataApi.list_data caps `limit` at 100 server-side.
        access_token=access_token,
    )
    return {Like.model_validate(doc).post_id for doc in result["data"]}


async def toggle_like(post_id: str, user_id: str, *, access_token: str) -> bool:
    """Returns the new liked state (True if now liked, False if now unliked).

    The like row itself (create/delete) is the operation that determines
    success — once it commits, the toggle has genuinely happened. The
    `likesCount` counter update on `posts` is a best-effort denormalization
    on top of that: if it fails (e.g. a transient Mudbase error), the failure
    is logged rather than raised, so a real like/unlike is never reported to
    the caller as failed just because the cosmetic counter didn't refresh.
    Mudbase has no atomic `$inc`, so a lost counter update here is a rare,
    self-correcting drift (the next real toggle on the same post reads the
    stale count and adjusts from there) rather than a compounding one.
    """
    settings = get_settings()
    existing = await find_like(post_id, user_id, access_token=access_token)
    if existing is not None:
        await asyncio.to_thread(delete_data_sync, settings.likes_collection_id, existing.id, access_token=access_token)
        await _adjust_likes_count_best_effort(post_id, delta=-1, access_token=access_token)
        return False

    await asyncio.to_thread(
        create_data_sync,
        settings.likes_collection_id,
        {"postId": post_id, "userId": user_id},
        access_token=access_token,
    )
    await _adjust_likes_count_best_effort(post_id, delta=1, access_token=access_token)
    return True


async def _adjust_likes_count_best_effort(post_id: str, *, delta: int, access_token: str) -> None:
    try:
        await adjust_post_counts(post_id, likes_delta=delta, access_token=access_token)
    except MudbaseApiError as exc:
        logger.warning(
            "Like toggle for post %s succeeded but the likesCount counter update failed: %s", post_id, exc.message
        )
